system-functionalities.md

```diff
@@ -338,14 +338,40 @@ Implementation guidance: For security, all verification checks run by the key se
 - ECDSA on secp256 curve; or
 - Ed25519 (i.e., EdDSA on edwards25519).
 
-Signing keys have an additional supported operation, namely, the creation of a signature.
+Signing keys have an additional supported operation, namely, the creation of a signature. They also have an additional public component that can be derived from a generated key.
 
 Implementation guidance: We pick these ECDSA/EdDSA parameters in order to provide functionality compatible with EVM-based blockchains. However, we expect to support multiple blockchains and signing primitives in the future.
 
+### Generating and Storing a Signing Key
+Both key generation protocols need a modification to return the public component of the generated key to the calling application.
+
+In [local secret generation with remote backup](#local-secret-generation-with-remote-encrypted-backup):
+- In step [3.iii](#localgen-client-generate), the generate protocol returns the `arbitrary_key` and its public component `public_key`.
+- In step [5.i](#localgen-client-output), the client outputs `key_id`, `public_key`, and `StorageObject` to the calling application.
+
+In [remote-only secret generation protocol](#remote-only-secret-generation-and-storage):
+- In step [2.iii](#remotegen-server-generate), the generate protocol returns the `arbitrary_key` and its public component `public_key`.
+- In step [2.v](#remotegen-server-send-key-id), the key server sends `key_id` and `public_key` to the client over the secure channel.
+- In step [3.i](#remotegen-client-store), the client stores `key_id` and `public_key`.
+- In [3.iii](#remotegen-client-output), the client outputs both `key_id` and `public_key` to the calling application.
+
+### Importing a Signing Key
+Both versions of key import should return the public component of the generated key to the calling application.
+
+In [local import with remote backup](#local-import-with-remote-backup):
+- In step [3](#localimport-client-encrypt), the client computes `public_key` from `arbitrary_key`. This can happen in any order relative to the other items in step 3.
+- In step [5.ii](#localimport-client-output), the client outputs both `key_id` and `public_key` to the calling application.
+
+In the [import to key server only protocol](#import-to-key-server-only):
+- In step [2.iii](#remoteimport-server-store), the server also computes `public_key` from `arbitrary_key`.
+- In step [2.iv](#remoteimport-server-send), the server sends both `key_id` and `public_key` to the client.
+- In step [3.i](#remoteimport-client-store), the client stores both the received `key_id` and `public_key`, and the associated context `"imported key"` locally.
+- In step [3.iii](#remoteimport-client-output), the client outputs both `key_id` and `public_key` to the calling application.
```

Comment:
missing validation checks on incoming messages. I should never have skimped on writing these, because I think there is a misconception that these checks aren't needed.

Reply:
Added these.